docs/source/advanced_patterns.rst

```diff
@@ -0,0 +1,107 @@
+
+An asynchronous request-reply channel is very similar to a normal request-reply
+pattern, only in this case we want to process the reply asynchronously (i.e.:
+assign a handler for this reply and forget about it until it is received).
+
+This means that, when using an asynchronous request-reply channel, a handler
+must be specified not only for the replier but also for the requester.
+
+Example
+-------
+
+See the following example in which Bob sends an asynchronous request to Alice.
+Note that:
+
+- When Alice binds, she uses ``ASYNC_REP`` instead of ``REP``.
+- Bob is assigned a handler too to process Alice reply when it is received.
+- Bob sends the request and is automatically freed. It can log messages or do
+  other stuff.
+- When the reply is received, Bob automatically processes it.
+
+.. literalinclude:: ../../examples/async_req_rep.py
+
+That is the most basic example, but channels are a bit more flexible. When
+sending an asynchronous request we may want to manually specify a handler for
+the reply. This will overwrite the default handler specified when connecting to
+the ``ASYNC_REP`` server. Try to change the ``.send()`` call::
+
+   def deaf(agent, message):
+       agent.log_info('I am deaf...')
+
+
+   bob.send('alice', 'Hello, Alice!', handler=deaf)
+
+We can also specify a maximum wait time for the reply and some code to be
+executed in case the reply was not received after that wait. Try the example
+above, but setting ``wait=0.5``::
+
+   def no_reply_in_time(agent):
+       agent.log_warning('No reply received!')
+
+
+   bob.send('alice', 'Hello, Alice!', wait=0.5, on_error=no_reply_in_time)
+
+.. note:: If ``on_error`` is not specified, by default the agent will simply
+   log a warning.
+
+
+.. index:: sync-pub-sub
+.. _sync_pub_sub:
+
+Synced Publish-Subscribe channel
+================================
+
+Another common use case is the synced publish-subscribe channel. It is similar
+to the normal publish-subscribe pattern but has some extra functionality. The
+server is configured to publish messages through a ``PUB`` socket and clients
+can connect to it to receive these messages. Using the synced publish-subscribe
+channel, however, allows the clients to send requests to the publisher.
+
+This can be really useful when we want to share some state/data from the
+publisher with all the subscribers but we want the subscribers to be able to
+interact (and perhaps even modify) this state/data.
+
+.. note:: Synchronization is possible because all replies are sent to the
+   requester through the same PUB-SUB socket as the one used for normal
+   publications. Note that, even though this reply is sent through the PUB-SUB
+   socket, it will only be received by the requester and not by any other
+   subscribed client.
+
+Example
+-------
+
+In the following example we have:
+
+- A synced publisher that requires a handler for requests when binding.
+- Two subscribers that connect to the publisher and subscribe to all topics.
```

Comment:
I think you are missing a "the" in "all the topics".

Reply:
It sounds better to me without the "the". And looking for search suggestions with Google...

![screenshot from 2018-02-07 10-36-01](https://user-images.githubusercontent.com/526577/35909067-b9dfb0aa-0bf2-11e8-814c-db8128a91ff1.png)

:stuck_out_tongue_winking_eye:

Reply:
Yeah, but in this case you are referring to all THE topics, meaning the topics you were talking about before, which is something the Google suggestions can't take into account. I'd rather add the "the", but you can leave it out if you want 😁

Reply:
As this is the first (and only) occurrence of "topics" in this documentation page, I'll leave it. :innocent: 

After all I'm saying "all topics" as in "any topic", not as "a list of topics" (i.e.: clients do not care about the topics the publisher might be using, they just want to subscribe to everything).